### ISSUE TITLE
HTTP endpoint - INSERT with ON CONFLICT DO UPDATE clause fails in unexpected way

### ISSUE THREAD

**Issue by dubiousdavid** [crate/crate#14906]
### CrateDB version

5.4.3

### CrateDB setup information

CrateDB Cloud.

### Problem description

INSERT with ON CONFLICT DO UPDATE clause fails in unexpected way.

### Steps to Reproduce

(1) Create table

```sql
CREATE TABLE IF NOT EXISTS "doc"."agencymeetingreadytoprocess" (
   "id" TEXT NOT NULL,
   "steps" OBJECT(STRICT) AS (
      "transcribe" TEXT,
      "parse" TEXT,
      "load" TEXT
   ),
   "timeline" OBJECT(STRICT) AS (
      "transcribe" TIMESTAMP WITH TIME ZONE,
      "parse" TIMESTAMP WITH TIME ZONE,
      "load" TIMESTAMP WITH TIME ZONE
   ),
   "transcription" OBJECT(STRICT) AS (
      "uniqueId" TEXT,
      "downloader" TEXT
   ),
   "meetingId" TEXT,
   "meetingDate" TIMESTAMP WITH TIME ZONE,
   "publishedDate" TIMESTAMP WITH TIME ZONE,
   "keywords" ARRAY(TEXT),
   "links" ARRAY(OBJECT(STRICT) AS (
      "url" TEXT
   )),
   "sourceType" TEXT,
   "sourceUrl" TEXT,
   "videoURLs" ARRAY(TEXT),
   "title" TEXT,
   "description" TEXT,
   "agency" OBJECT(STRICT) AS (
      "agencyId" BIGINT,
      "name" TEXT,
      "stateCode" TEXT
   ),
   "documents" ARRAY(OBJECT(STRICT) AS (
      "url" TEXT,
      "requestURL" TEXT,
      "type" TEXT,
      "contentType" TEXT,
      "contentLength" BIGINT,
      "filename" TEXT
   )),
   "createdAt" TIMESTAMP WITH TIME ZONE,
   "updatedAt" TIMESTAMP WITH TIME ZONE,
   PRIMARY KEY ("id")
)
CLUSTERED BY ("id") INTO 6 SHARDS
WITH (
   "allocation.max_retries" = 5,
   "blocks.metadata" = false,
   "blocks.read" = false,
   "blocks.read_only" = false,
   "blocks.read_only_allow_delete" = false,
   "blocks.write" = false,
   codec = 'default',
   column_policy = 'strict',
   "mapping.total_fields.limit" = 1000,
   max_ngram_diff = 1,
   max_shingle_diff = 3,
   number_of_replicas = '0-1',
   "routing.allocation.enable" = 'all',
   "routing.allocation.total_shards_per_node" = -1,
   "store.type" = 'fs',
   "translog.durability" = 'REQUEST',
   "translog.flush_threshold_size" = 536870912,
   "translog.sync_interval" = 5000,
   "unassigned.node_left.delayed_timeout" = 60000,
   "write.wait_for_active_shards" = '1'
)
```

(2) Run the following two times, replacing auth with the basic auth string.

```typescript
import fetch from 'node-fetch'

const auth = '...'

const getAuthHeader = (auth: string) => ({
  Authorization: 'Basic ' + Buffer.from(auth, 'binary').toString('base64'),
})
const sqlEndpoint =
  'http://localhost:4200/_sql'
const sql = `INSERT INTO "doc"."agencymeetingreadytoprocess" ("meetingId","agency","createdAt","meetingDate","sourceType","sourceUrl","steps","title","updatedAt","videoURLs","timeline","transcription","id") VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)
    ON CONFLICT (id) DO UPDATE SET "updatedAt" = ?,"timeline" = ?,"transcription" = ?,"steps"['transcribe'] = ?`

const args = [
  'swagit-coronadousdca-10102023-517',
  {
    agencyId: 72611,
    name: 'Coronado Unified School District',
    stateCode: 'CA',
  },
  '2023-10-10T20:26:19.172Z',
  '2023-10-05T12:00:00.000Z',
  'swagit',
  'http://coronadousdca.swagit.com/specialty',
  { load: 'complete', parse: 'incomplete', transcribe: 'complete' },
  'KCMS 10-5 Broadcast',
  '2023-10-10T21:20:53.694Z',
  [
    'https://admin-v2.swagit.com/podcasts/2023/10/10/10102023-518.360.mp4',
    'https://stream.us-central1-b.swagit.com/vod/_definst_/admin-v2/2023/10/10/10102023-518.360.mp4/playlist.m3u8',
  ],
  {
    transcribe: '2023-10-10T21:10:00.642Z',
    load: '2023-10-10T21:20:53.694Z',
  },
  { uniqueId: '6525b448c484e452fe48df08' },
  '6525b36b3d9cfcf8ec8f732b',
  '2023-10-10T20:30:00.748Z',
  { transcribe: '2023-10-10T20:30:00.748Z' },
  { uniqueId: '6525b448c484e452fe48df08' },
  'pending',
]

const result = await fetch(sqlEndpoint, {
  method: 'post',
  body: JSON.stringify({ stmt: sql, args }),
  headers: { 'Content-Type': 'application/json', ...getAuthHeader(auth) },
})

console.dir(await result.json(), { depth: 10 })
```

(3) Observe the following error

```json
{
  "error": {
    "message": "SQLParseException[Cannot add column `transcribe` to strict object `agency`]",
    "code": 4000
  }
}
```

### Actual Result

```json
{
  "error": {
    "message": "SQLParseException[Cannot add column `transcribe` to strict object `agency`]",
    "code": 4000
  }
}
```

### Expected Result

Success.

**Comment by seut** [comment]
Looks like you are using a wrong JSON payload schema. As stated here https://cratedb.com/docs/crate/reference/en/5.4/interfaces/http.html, the `args` must be set as the value of an `args` key.
So the `body` should be something like:
```
body: JSON.stringify({ "stmt": sql, "args": args }),
```

**Comment by dubiousdavid** [comment]
That is just object property shorthand syntax. It's equivalent to `"args": args`.

**Comment by dubiousdavid** [comment]
This code is part of a library that has been successfully running in production for a year. My guess is that there is a regression in CrateDB. https://www.npmjs.com/package/mongo2crate

**Comment by amotl** [comment]
It is just a wild guess, without running the example yet: Maybe the notation within the `DO UPDATE` clause is slightly off at the very end: Is it correct to address the subcolumn like `"steps"['transcribe']`, or would `"steps['transcribe']"` be more appropriate?

**Comment by seut** [comment]
> It is just a wild guess, without running the example yet: Maybe the notation within the `DO UPDATE` clause is slightly off at the very end: Is it correct to address the subcolumn like `"steps"['transcribe']`, or would `"steps['transcribe']"` be more appropriate?

Yes it is, and no `"steps['transcribe']"` is wrong.

**Comment by amotl** [comment]
I did not investigate about anything else yet, but the plain SQL variant, without using statement parameters, works well.

- [testdrive-agencymeeting.sql.txt](https://github.com/crate/crate/files/12937570/testdrive-agencymeeting.sql.txt)

**Comment by seut** [comment]
@dubiousdavid I could reproduce the issue, thank you for reporting.

**Comment by amotl** [comment]
A corresponding Python program fails with the same error.

```
wget -O testdrive-agencymeeting.py https://gist.github.com/amotl/31701b4d4162090e9815f4c209b57cb9
```

```python
$ python testdrive-agencymeeting.py
Traceback (most recent call last):
  File "/Users/amo/dev/tmp/testdrive/testdrive-agencymeeting.py", line 46, in <module>
    main()
  File "/Users/amo/dev/tmp/testdrive/testdrive-agencymeeting.py", line 42, in main
    cursor.execute(sql, args)
  File "/Users/amo/dev/tmp/testdrive/.venv/lib/python3.9/site-packages/crate/client/cursor.py", line 58, in execute
    self._result = self.connection.client.sql(sql, parameters,
  File "/Users/amo/dev/tmp/testdrive/.venv/lib/python3.9/site-packages/crate/client/http.py", line 454, in sql
    content = self._json_request('POST', self.path, data=data)
  File "/Users/amo/dev/tmp/testdrive/.venv/lib/python3.9/site-packages/crate/client/http.py", line 581, in _json_request
    _raise_for_status(response)
  File "/Users/amo/dev/tmp/testdrive/.venv/lib/python3.9/site-packages/crate/client/http.py", line 199, in _raise_for_status
    return _raise_for_status_real(response)
  File "/Users/amo/dev/tmp/testdrive/.venv/lib/python3.9/site-packages/crate/client/http.py", line 228, in _raise_for_status_real
    raise ProgrammingError(error.get('message', ''),
crate.client.exceptions.ProgrammingError: SQLParseException[Cannot add column `transcribe` to strict object `agency`]
```

The reason is the very last parameter to the `ON CONFLICT` clause. When omitting `"steps"['transcribe'] = ?` and `pending`, it works.

**Comment by hlcianfagna** [comment]
Hi @dubiousdavid
I think the problem can be worked around by using this expression:
```sql
ON CONFLICT (id) DO UPDATE SET "updatedAt" = ?,"timeline" = ?,"transcription" = ?,
"steps" = concat(excluded.steps, concat('{"transcribe":"',?,'"}'))
```
Please give it a try and let us know if that works for you.
Thank you.

**Comment by seut** [comment]
@dubiousdavid
We've fixed the issue by #14931, it will be part of the next `5.4.5` hotfix release. 
It was indeed a regression introduced by CrateDB `5.3.0`.
Thank you again for reporting.

**Comment by dubiousdavid** [comment]
> @dubiousdavid We've fixed the issue by #14931, it will be part of the next `5.4.5` hotfix release. It was indeed a regression introduced by CrateDB `5.3.0`. Thank you again for reporting.

Thank you!
